File: android/app/src/main/java/com/ayuvo/health/insights/InsightsMath.kt
package com.ayuvo.health.insights

import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.temporal.ChronoUnit
import java.util.Locale
import kotlin.math.abs
import kotlin.math.floor
import kotlin.math.sqrt

/**
 * Numbers, days and text shared by every Insights engine, written out exactly as in
 * `scripts/insights_reference.py` so Kotlin and Swift doubles give identical results
 * (docs/insights.md §2.2): half-up rounding by `floor(x·10^d + 0.5)`, chronological sums,
 * sample SD with n − 1, U+2212 for minus signs.
 */
object InsightsMath {
    const val MINUS = "−"
    private val PLACEHOLDER = Regex("[{]([a-z_][a-z0-9_]*)[}]")
    private val SCALES = doubleArrayOf(1.0, 10.0, 100.0, 1000.0, 10000.0)

    fun roundTo(x: Double, decimals: Int): Double {
        val scale = SCALES[decimals]
        val v = floor(x * scale + 0.5) / scale
        return if (v == 0.0) 0.0 else v // never -0.0
    }

    fun roundTo(x: Double?, decimals: Int): Double? = x?.let { roundTo(it, decimals) }

    fun roundInt(x: Double): Int = floor(x + 0.5).toInt()

    fun clamp(x: Double, lo: Double, hi: Double): Double = if (x < lo) lo else if (x > hi) hi else x

    fun mean(values: List<Double>): Double {
        var total = 0.0
        for (v in values) total += v
        return total / values.size
    }

    /** Sample standard deviation (divides by n − 1); needs n ≥ 2. */
    fun sampleSd(values: List<Double>, m: Double): Double {
        var total = 0.0
        for (v in values) total += (v - m) * (v - m)
        return sqrt(total / (values.size - 1))
    }

    /** Least-squares slope of `(x, y)`; null when x has no spread. */
    fun olsSlope(points: List<Pair<Double, Double>>): Double? {
        val mx = mean(points.map { it.first })
        val my = mean(points.map { it.second })
        var sxx = 0.0
        var sxy = 0.0
        for ((x, y) in points) {
            sxx += (x - mx) * (x - mx)
            sxy += (x - mx) * (y - my)
        }
        return if (sxx == 0.0) null else sxy / sxx
    }

    /** `b − a` in whole days. */
    fun daysBetween(a: LocalDate, b: LocalDate): Long = ChronoUnit.DAYS.between(a, b)

    fun localDayOf(ms: Long, zone: ZoneId): LocalDate = Instant.ofEpochMilli(ms).atZone(zone).toLocalDate()

    fun localMinutesOf(ms: Long, zone: ZoneId): Int {
        val t = Instant.ofEpochMilli(ms).atZone(zone)
        return t.hour * 60 + t.minute
    }

    /**
     * Wall-clock minutes from 12:00 on the day before [wakeDay] to the night's midpoint: a
     * 23:00–07:00 night gives 900 (the same clock as the metrics `sleep_clock_offset`).
     */
    fun sleepMidpointMin(night: SleepInput, wakeDay: LocalDate, zone: ZoneId): Int {
        val mid = night.startMs + Math.floorDiv(night.endMs - night.startMs, 2L)
        val local = Instant.ofEpochMilli(mid).atZone(zone)
        val days = ChronoUnit.DAYS.between(wakeDay.minusDays(1), local.toLocalDate()).toInt()
        return days * 1440 + local.hour * 60 + local.minute - 720
    }

    fun groupInt(n: Long): String {
        var s = abs(n).toString()
        val out = ArrayList<String>()
        while (s.length > 3) {
            out.add(0, s.substring(s.length - 3))
            s = s.substring(0, s.length - 3)
        }
        out.add(0, s)
        return (if (n < 0) "-" else "") + out.joinToString(",")
    }

    /** Template number: integral → grouped integer ("8,432"), otherwise one decimal ("7.5"). */
    fun fmtNumber(x: Double): String =
        if (x == floor(x)) groupInt(x.toLong()) else String.format(Locale.ROOT, "%.1f", roundTo(x, 1))

    fun fmtSigned(x: Double, decimals: Int): String {
        val r = roundTo(x, decimals)
        if (r == 0.0) return "0"
        val body = if (decimals == 0) groupInt(abs(r).toLong()) else String.format(Locale.ROOT, "%.${decimals}f", abs(r))
        return (if (r > 0) "+" else MINUS) + body
    }

    fun fmtDuration(minutes: Double): String {
        val m = roundInt(minutes)
        return if (m < 60) "${m}m" else "${m / 60}h ${m % 60}m"
    }

    /** Replaces each `{name}` with its param: numbers via [fmtNumber], strings as they are. */
    fun fill(template: String, params: Map<String, Any?>): String =
        PLACEHOLDER.replace(template) { m ->
            when (val v = params[m.groupValues[1]] ?: error("missing param ${m.groupValues[1]}")) {
                is String -> v
                is Number -> fmtNumber(v.toDouble())
                else -> v.toString()
            }
        }

    /** Piecewise-linear y(x) over `[x, y]` points with x ascending; flat beyond the ends. */
    fun interp(points: Points, x: Double): Double {
        if (x <= points[0].first) return points[0].second
        for (i in 0 until points.size - 1) {
            val (x0, y0) = points[i]
            val (x1, y1) = points[i + 1]
            if (x <= x1) return y0 + (y1 - y0) * (x - x0) / (x1 - x0)
        }
        return points.last().second
    }

    /**
     * Age at which a monotonic population curve `[age, value]` equals [value]: inverse linear
     * interpolation, linear extrapolation from the end segments, then clamped to the age range.
     */
    fun inverseAge(table: Points, value: Double, ageMin: Double, ageMax: Double): Double {
        val n = table.size
        var seg = -1
        for (i in 0 until n - 1) {
            val lo = minOf(table[i].second, table[i + 1].second)
            val hi = maxOf(table[i].second, table[i + 1].second)
            if (value in lo..hi) {
                seg = i
                break
            }
        }
        if (seg < 0) {
            val increasing = table[n - 1].second > table[0].second
            val belowFirst = if (increasing) value < table[0].second else value > table[0].second
            seg = if (belowFirst) 0 else n - 2
        }
        val (a0, v0) = table[seg]
        val (a1, v1) = table[seg + 1]
        val age = a0 + (value - v0) * (a1 - a0) / (v1 - v0)
        return clamp(age, ageMin, ageMax)
    }

    /** `(offset, value)` for days `day + first .. day + last` (inclusive, chronological), present values only. */
    fun values(series: Map<LocalDate, Double>, day: LocalDate, first: Int, last: Int): List<Pair<Int, Double>> {
        val out = ArrayList<Pair<Int, Double>>()
        for (k in first..last) series[day.plusDays(k.toLong())]?.let { out += k to it }
        return out
    }
}

File: ios/calorietracker/Insights/InsightsMath.swift
import Foundation

// Numbers, days and text shared by every Insights engine. Line-by-line port of the helpers at the top of
// `scripts/insights_reference.py` (docs/insights.md §2.2): the reference wins over prose.

nonisolated enum InsightsMath {
    static let minus = "\u{2212}"

    /// `floor(x * 10^d + 0.5) / 10^d`, never -0.
    static func roundTo(_ x: Double, _ decimals: Int) -> Double {
        let scale: Double
        switch decimals {
        case 0: scale = 1
        case 1: scale = 10
        case 2: scale = 100
        default: scale = pow(10, Double(decimals))
        }
        let v = (x * scale + 0.5).rounded(.down) / scale
        return v == 0 ? 0 : v
    }

    @_disfavoredOverload
    static func roundTo(_ x: Double?, _ decimals: Int) -> Double? {
        x.map { roundTo($0, decimals) }
    }

    static func roundInt(_ x: Double) -> Int { Int((x + 0.5).rounded(.down)) }

    @_disfavoredOverload
    static func roundInt(_ x: Double?) -> Int? { x.map { roundInt($0) } }

    static func clamp(_ x: Double, _ lo: Double, _ hi: Double) -> Double {
        x < lo ? lo : (x > hi ? hi : x)
    }

    /// Chronological / list-order sum divided by n.
    static func mean(_ values: [Double]) -> Double {
        var total = 0.0
        for v in values { total += v }
        return total / Double(values.count)
    }

    /// Sample standard deviation (n − 1). Needs n ≥ 2.
    static func sampleSD(_ values: [Double], _ m: Double) -> Double {
        var total = 0.0
        for v in values { total += (v - m) * (v - m) }
        return (total / Double(values.count - 1)).squareRoot()
    }

    /// Least-squares slope of (x, y); nil when x has no spread.
    static func olsSlope(_ points: [(Double, Double)]) -> Double? {
        let mx = mean(points.map(\.0))
        let my = mean(points.map(\.1))
        var sxx = 0.0, sxy = 0.0
        for (x, y) in points {
            sxx += (x - mx) * (x - mx)
            sxy += (x - mx) * (y - my)
        }
        return sxx == 0 ? nil : sxy / sxx
    }

    /// Piecewise-linear y(x) over [[x, y]] with x ascending; flat beyond the ends.
    static func interp(_ points: [[Double]], _ x: Double) -> Double {
        if x <= points[0][0] { return points[0][1] }
        for i in 0..<(points.count - 1) {
            let (x0, y0, x1, y1) = (points[i][0], points[i][1], points[i + 1][0], points[i + 1][1])
            if x <= x1 { return y0 + (y1 - y0) * (x - x0) / (x1 - x0) }
        }
        return points[points.count - 1][1]
    }

    /// Age at which a monotonic population curve [[age, value]] equals `value` (inverse linear interpolation,
    /// linear extrapolation from the end segments), clamped to [ageMin, ageMax].
    static func inverseAge(_ table: [[Double]], _ value: Double, _ ageMin: Double, _ ageMax: Double) -> Double {
        let n = table.count
        var seg = n - 2
        var found = false
        for i in 0..<(n - 1) {
            let lo = min(table[i][1], table[i + 1][1]), hi = max(table[i][1], table[i + 1][1])
            if lo <= value, value <= hi {
                seg = i
                found = true
                break
            }
        }
        if !found {
            let increasing = table[n - 1][1] > table[0][1]
            let belowFirst = increasing ? value < table[0][1] : value > table[0][1]
            seg = belowFirst ? 0 : n - 2
        }
        let (a0, v0, a1, v1) = (table[seg][0], table[seg][1], table[seg + 1][0], table[seg + 1][1])
        let age = a0 + (value - v0) * (a1 - a0) / (v1 - v0)
        return clamp(age, ageMin, ageMax)
    }

    static func floorDiv(_ a: Int64, _ b: Int64) -> Int64 {
        let q = a / b
        return (a % b != 0 && (a < 0) != (b < 0)) ? q - 1 : q
    }

    static func floorMod(_ a: Int64, _ b: Int64) -> Int64 { a - floorDiv(a, b) * b }
}

// MARK: - Days ("YYYY-MM-DD")

nonisolated enum InsightsDay {
    /// Days since 1970-01-01 for a civil date (proleptic Gregorian).
    static func ordinal(year: Int, month: Int, day: Int) -> Int {
        let y = month <= 2 ? year - 1 : year
        let era = (y >= 0 ? y : y - 399) / 400
        let yoe = y - era * 400
        let mp = (month + 9) % 12
        let doy = (153 * mp + 2) / 5 + day - 1
        let doe = yoe * 365 + yoe / 4 - yoe / 100 + doy
        return era * 146_097 + doe - 719_468
    }

    static func civil(_ ordinal: Int) -> (year: Int, month: Int, day: Int) {
        let z = ordinal + 719_468
        let era = (z >= 0 ? z : z - 146_096) / 146_097
        let doe = z - era * 146_097
        let yoe = (doe - doe / 1460 + doe / 36524 - doe / 146_096) / 365
        let doy = doe - (365 * yoe + yoe / 4 - yoe / 100)
        let mp = (5 * doy + 2) / 153
        let d = doy - (153 * mp + 2) / 5 + 1
        let m = mp < 10 ? mp + 3 : mp - 9
        return (yoe + era * 400 + (m <= 2 ? 1 : 0), m, d)
    }

    /// Ordinal of a "YYYY-MM-DD" key; nil when malformed.
    static func parse(_ s: String) -> Int? {
        let u = Array(s.utf8)
        guard u.count == 10, u[4] == 45, u[7] == 45 else { return nil }
        func num(_ r: Range<Int>) -> Int? {
            var v = 0
            for i in r {
                guard u[i] >= 48, u[i] <= 57 else { return nil }
                v = v * 10 + Int(u[i] - 48)
            }
            return v
        }
        guard let y = num(0..<4), let m = num(5..<7), let d = num(8..<10), (1...12).contains(m), (1...31).contains(d)
        else { return nil }
        return ordinal(year: y, month: m, day: d)
    }

    static func format(_ ordinal: Int) -> String {
        let c = civil(ordinal)
        return String(format: "%04d-%02d-%02d", c.year, c.month, c.day)
    }

    static func add(_ day: String, _ n: Int) -> String {
        guard let o = parse(day) else { return day }
        return format(o + n)
    }

    /// b − a in whole days.
    static func between(_ a: String, _ b: String) -> Int {
        (parse(b) ?? 0) - (parse(a) ?? 0)
    }

    static func timeZone(_ id: String) -> TimeZone {
        TimeZone(identifier: id) ?? TimeZone(identifier: "UTC")!
    }

    private static func localMs(_ ms: Int64, _ tz: TimeZone) -> Int64 {
        let date = Date(timeIntervalSince1970: Double(ms) / 1000.0)
        return ms + Int64(tz.secondsFromGMT(for: date)) * 1000
    }

    static func localDayOrdinal(ms: Int64, _ tz: TimeZone) -> Int {
        Int(InsightsMath.floorDiv(localMs(ms, tz), 86_400_000))
    }

    static func localDay(ms: Int64, _ tz: TimeZone) -> String { format(localDayOrdinal(ms: ms, tz)) }

    /// Wall-clock minutes since local midnight.
    static func localMinutes(ms: Int64, _ tz: TimeZone) -> Int {
        Int(InsightsMath.floorMod(localMs(ms, tz), 86_400_000) / 60_000)
    }

    /// Wall-clock minutes from 12:00 on the day before `wakeDay` to the night's midpoint: a 23:00–07:00 night → 900.
    static func sleepMidpointMinutes(startMs: Int64, endMs: Int64, wakeDay: String, _ tz: TimeZone) -> Int {
        let mid = startMs + InsightsMath.floorDiv(endMs - startMs, 2)
        let days = localDayOrdinal(ms: mid, tz) - ((parse(wakeDay) ?? 0) - 1)
        return days * 1440 + localMinutes(ms: mid, tz) - 720
    }

    /// Local day key of `date` in `calendar`'s time zone.
    static func key(for date: Date, calendar: Calendar = .current) -> String {
        let c = calendar.dateComponents([.year, .month, .day], from: date)
        return format(ordinal(year: c.year ?? 1970, month: c.month ?? 1, day: c.day ?? 1))
    }

    /// Local noon of a day key, for display.
    static func date(_ day: String, calendar: Calendar = .current) -> Date? {
        guard let o = parse(day) else { return nil }
        let c = civil(o)
        return calendar.date(from: DateComponents(year: c.year, month: c.month, day: c.day, hour: 12))
    }
}

// MARK: - Text

nonisolated enum InsightsFormat {
    static func groupInt(_ n: Int) -> String {
        var s = String(n.magnitude)
        var out: [String] = []
        while s.count > 3 {
            out.insert(String(s.suffix(3)), at: 0)
            s = String(s.dropLast(3))
        }
        out.insert(s, at: 0)
        return (n < 0 ? "-" : "") + out.joined(separator: ",")
    }

    /// Integral → grouped integer ("8,432"), otherwise one decimal ("7.5").
    static func number(_ x: Double) -> String {
        if x == x.rounded(.down) { return groupInt(Int(x)) }
        return String(format: "%.1f", InsightsMath.roundTo(x, 1))
    }

    /// "+8", "−3", "+0.4" or "0".
    static func signed(_ x: Double, _ decimals: Int) -> String {
        let r = InsightsMath.roundTo(x, decimals)
        if r == 0 { return "0" }
        let body = decimals == 0 ? groupInt(Int(abs(r))) : String(format: "%.\(decimals)f", abs(r))
        return (r > 0 ? "+" : InsightsMath.minus) + body
    }

    /// "7h 48m" or "45m".
    static func duration(_ minutes: Double) -> String {
        let m = InsightsMath.roundInt(minutes)
        return m < 60 ? "\(m)m" : "\(m / 60)h \(m % 60)m"
    }

    /// Replaces each `{name}` with its param: numbers via `number`, strings as is.
    static func fill(_ template: String, _ params: [String: InsightsParam]) -> String {
        var out = ""
        var rest = Substring(template)
        while let open = rest.firstIndex(of: "{") {
            out += rest[rest.startIndex..<open]
            let after = rest.index(after: open)
            guard let close = rest[after...].firstIndex(of: "}") else {
                rest = rest[open...]
                break
            }
            let name = rest[after..<close]
            if isPlaceholderName(name), let value = params[String(name)] {
                out += value.templateText
                rest = rest[rest.index(after: close)...]
            } else {
                out += "{"
                rest = rest[after...]
            }
        }
        out += rest
        return out
    }

    private static func isPlaceholderName(_ s: Substring) -> Bool {
        guard let first = s.unicodeScalars.first, first == "_" || ("a"..."z").contains(first) else { return false }
        return s.unicodeScalars.allSatisfy { $0 == "_" || ("a"..."z").contains($0) || ("0"..."9").contains($0) }
    }
}
